### PR TITLE
[Plugin] Add APIs for track segments and track iteration

### PR DESCRIPTION
**Pull request by IntelOrca**
```ts
interface Context {
    getTrackSegment(type: number): TrackSegment | null;
}

interface GameMap {
    getTrackIterator(location: CoordsXY, elementIndex: number): TrackIterator | null;
}

interface TrackSegment {
    readonly type: number;
    readonly description: string;
    readonly beginZ: number;
    readonly beginDirection: Direction8;
    readonly endX: number;
    readonly endY: number;
    readonly endZ: number;
    readonly endDirection: Direction8;
    readonly elements: TrackSegmentElement[];
}

interface TrackSegmentElement implements CoordsXYZ {
}

interface TrackIterator {
    readonly position: CoordsXYZD;
    readonly segment: TrackSegment | null;
    readonly previousPosition: CoordsXYZD | null;
    readonly nextPosition: CoordsXYZD | null;

    previous(): boolean;
    next(): boolean;
}
```

Demo script:
https://gist.github.com/IntelOrca/680cbcd9d0df497396a342f155c51691

### CODE REVIEW

**Comment by oli414** [comment]
The API looks good to me

**Comment by Basssiiie** [comment]
Hey, looks good, Intel! Some questions:
- Will it be possible to get a TrackElement directly from a TrackSegmentElement/TrackIterator? This would prevent issues where the game and plugin pick inconsistent elements when they are at the same xyz coordinates.
- Will track type id always be a number or are there plans to eventually move to track piece objects?
- Would it be possible to have some sort of length added to the TrackSegment?

Thank you for your time!

**Comment by duncanspumpkin** [comment]
How will users get the element index. Shouldn't the interface be an xyzd, rideid. The element index isn't exactly a great thing to pass around in plugin code as plugins could get called with ghosts placed which will obviously mess up any indexes.

**Comment by IntelOrca** [comment]
@duncanspumpkin I choose element index because Z is always going to be ambiguous when there is track merging, crossing elements etc.

> Will it be possible to get a TrackElement directly from a TrackSegmentElement/TrackIterator? This would prevent issues where the game and plugin pick inconsistent elements when they are at the same xyz coordinates.

I think this could differ in different places within the code base. I decided to let the plugin writer decide which element they are going to select if multiple exist at the same location.


> Will track type id always be a number or are there plans to eventually move to track piece objects?

We won't have objects for track segments, but even if we do - you can assume the track type ID is an object ID, and the getTrackSegment is how you get the information for that segment, given the object ID.

> Would it be possible to have some sort of length added to the TrackSegment?

Sure, will add that.

> Edit: I also recommend returning null instead of undefined if something is missing. It is more explicit.

I have decided to use undefined exclusively now, just because TypeScript uses undefined for anything that is optional. It appears that in vscode, Microsoft use undefined for all their APIs too.
https://stackoverflow.com/questions/49746803/typescript-null-versus-undefined

**Comment by duncanspumpkin** [comment]
> @duncanspumpkin I choose element index because Z is always going to be ambiguous when there is track merging, crossing elements etc.
> 
> I could add another overload though for `[ride id, xyzd]`.

As long as plugin developers understand that the element index is ambiguous across hooks. Technically xyzd rideid isn't ambiguous it always returns the first match just unfortunately that might not be what you want.

**Comment by IntelOrca** [comment]
@duncanspumpkin You would also need to give it the track type as well. Basically it needs x, y, z, direction, track type and ride ID. which is a lot of data in comparison to the element index if you already have that. So I think it would be useful to have both. It probably doesn't matter if it is ambiguous, because the iterator would give the same results if those inputs were the same.

**Comment by Basssiiie** [comment]
> I think this could differ in different places within the code base. I decided to let the plugin writer decide which element they are going to select if multiple exist at the same location.

@IntelOrca I'd say the most common use case is to follow OpenRCT2's _preferred_ way of selecting the right tile element, so a quick way to get that tile element would IMHO be best instead of plugins all try to mimic one of the various OpenRCT2's ways. That's bound to create many plugin inconsistencies.

And if the plugin developer still really wants to do it his or her own way, they can of course still do it with the xyzd/type/ride/banking/angle/inverted or not/covered dinghy track or not/whatever else is involved in selecting the right one. :)

> I have decided to use undefined exclusively now, just because TypeScript uses undefined for anything that is optional. It appears that in vscode, Microsoft use undefined for all their APIs too.

The preference is taken from [this page](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines), which opens with:

> STOP READING IMMEDIATELY
> THIS PAGE PROBABLY DOES NOT PERTAIN TO YOU.
> These are Coding Guidelines for Contributors to TypeScript. This is NOT a prescriptive guideline for the TypeScript community. > These guidelines are meant for contributors to the TypeScript project's codebase. We have chosen many of them for team consistency. Feel free to adopt them for your own team.
> AGAIN: This is NOT a prescriptive guideline for the TypeScript community

Furthermore, the Typescript API's just mimic whatever the JS API's were originally designed to return. Though I personally think 'null' would be better and more explicit, but it seems mostly personal preference.

**Comment by Sadret** [comment]
A little bit late to the discussion, but these are my thoughts:

When I heard of 'track iteration', I thought that you would create an iterator by `ride id`, with no other params at all. For me that would probably be the most common use case.
The other use case would be to get the iterator by some `TrackElement`. I.e.. using a `Tool` to select some track element with your cursor and receive an iterator of the selected ride, which is at the position of the current track piece. So with the current implementation, at what position is the `TrackIterator` that is returned by `getTrackIterator`?

As for `TrackSegment.elements`, it should definitely provide a unique identification of the elements. I would choose `mapCoords` + `tileElementIndex`, similiar to what you have in `ToolEventArgs`.

A question regarding `TrackSegment.endX` and alike: is it the position relative to the origin? I.e.
`iter.position.x + iter.segment.endX == iter.next.position.x`.

**Comment by IntelOrca** [comment]
@Basssiiie I remember why I choose undefined now. It all came down to the behaviour of [optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html).

`obj?.field` in TypeScript equates to `p === null || p === undefined ? undefined : p.field`. And if you want consistency in using one or the other throughout your code, this operator kind of puts you down the undefined road.

**Comment by Basssiiie** [comment]
@IntelOrca Isn't that because `p.field` will also return `undefined` if the `field` key does not exist on `p`? In this case it's not a missing key though, it's a missing value.

**Comment by IntelOrca** [comment]
@Basssiiie if you look at this: https://code.visualstudio.com/api/references/vscode-api

They use `undefined` all the time as a return value.

**Comment by Basssiiie** [comment]
@IntelOrca You're right, but they still use `null` as well. They do note this [here](https://code.visualstudio.com/api/references/vscode-api#strict-null):

> The VS Code API uses the `undefined` and `null` TypeScript types where appropriate

But I couldn't find their official argumentation on what is appropriate in their code base/API.

Furthermore, I also mentioned...

> but it seems mostly personal preference.

so please do not feel forced to argue it. 😋 (unless you want to discuss it more)

**Comment by Basssiiie** [comment]
Excuse me for another message, but one thing I noticed today that the plugin API is already using `null` for missing elements in other places, e.g. [here](https://github.com/OpenRCT2/OpenRCT2/blob/563a893dfe2987b65a3d40ae98398c80bbce1c05/distribution/openrct2.d.ts#L1211-L1227):
```ts
 interface Car extends Entity {
        nextCarOnTrain: number | null;
        previousCarOnRide: number | null;
        nextCarOnRide: number | null;
}
```
So using `undefined` would make this implementation inconsistent with the rest of the API.

**Comment by IntelOrca** [comment]
@Basssiiie A lot of the graphics APIs are using `undefined`.

**Comment by Basssiiie** [comment]
@IntelOrca That is true, a quick ctrl+f search through openrct2.d.ts seems to show that the Graphics API was one of the first to deviate from the `null` consistency.

- `undefined` appears only 9 times (excluding comments) in openrct2.d.ts.
- `null` appears 34 times (excluding comments) in openrct2.d.ts.

Not saying that `undefined` should never be used, but there's definitely some mixing of different styles.

**Comment by IntelOrca** [comment]
@Basssiiie I have changed this PR to returning null.

File: distribution/openrct2.d.ts [hunk 1]
```diff
@@ -1128,6 +1144,138 @@ declare global {
...
+        readonly endX: number;
+
+        /**
+         * The relative ending Y position.
+         */
+        readonly endY: number;
+
+        /**
+         * The relative ending Z position. Negative numbers indicate
+         * that the track ends upside down.
```
Comment (Broxzier) [inline distribution/openrct2.d.ts:1191]: With 'upside down' you mean an inversion?

**Review by Broxzier — APPROVED** [review]
Tested it with a quick plug-in with two buttons to iterate manually, as well as your test script. Everything seems to behave as expected, and I didn't see any parts of the code that need changing.

File: distribution/openrct2.d.ts (hunk 1 above)
Comment (IntelOrca) [reply distribution/openrct2.d.ts:1191]: I think so? Presumably the half corkscrew or the half loop track pieces...

**Pull request merged**